File: apps/web/src/features/effects/featureFlags.ts
export type EffectFeatureFlags = {
  persistence: boolean;
  preview: boolean;
  render: boolean;
};

export function effectFeatureFlags(env: Record<string, string | undefined>): EffectFeatureFlags {
  const read = (key: string) => ['1', 'true', 'yes', 'on'].includes((env[key] ?? '').toLowerCase());
  const flags = {
    persistence: read('VITE_EFFECT_V2_PERSISTENCE'),
    preview: read('VITE_EFFECT_V2_PREVIEW'),
    render: read('VITE_EFFECT_V2_RENDER'),
  };
  if ((flags.preview || flags.render) && !flags.persistence) {
    throw new Error('Effect Engine V2 preview/render requires persistence');
  }
  return flags;
}

File: apps/api/src/workbench/effects/feature_flags.py
from __future__ import annotations

import os
from dataclasses import dataclass


def _enabled(name: str, default: bool = False) -> bool:
    return os.environ.get(name, str(default)).strip().lower() in {"1", "true", "yes", "on"}


@dataclass(frozen=True)
class EffectFeatureFlags:
    persistence: bool = False
    preview: bool = False
    render: bool = False

    @classmethod
    def from_environment(cls) -> EffectFeatureFlags:
        flags = cls(
            persistence=_enabled("WORKBENCH_EFFECT_V2_PERSISTENCE"),
            preview=_enabled("WORKBENCH_EFFECT_V2_PREVIEW"),
            render=_enabled("WORKBENCH_EFFECT_V2_RENDER"),
        )
        if (flags.preview or flags.render) and not flags.persistence:
            raise ValueError("Effect Engine V2 preview/render requires persistence")
        return flags
